### PR TITLE
HTTPClient 4.5. migration to last APIs / Bugzilla 58757 & 61664

### PR DESCRIPTION
**Pull request by ubikloadpack**
Hello,
Find in this PR the migration to last HC4 APIs and the fix of Digest Auth that was only partially implemented.

This PR fixes:
- https://bz.apache.org/bugzilla/show_bug.cgi?id=61664 
- https://bz.apache.org/bugzilla/show_bug.cgi?id=58757

Regards
Philippe M.

### CODE REVIEW

**Comment by pmouawad** [comment]
Hello,
Any feedback on this ?

Thanks

File: src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPHC4Impl.java [hunk 1]
```diff
@@ -215,22 +269,128 @@ public long getKeepAliveDuration(HttpResponse response, HttpContext context) {
...
-        HttpConnectionMetrics metrics = ((HttpConnection) context.getAttribute(HttpCoreContext.HTTP_CONNECTION)).getMetrics();
-        context.setAttribute(CONTEXT_METRICS, metrics);
-    };
-    private static final HttpRequestInterceptor METRICS_RESETTER = (HttpRequest request, HttpContext context) -> {
-        HttpConnectionMetrics metrics = ((HttpConnection) context.getAttribute(HttpCoreContext.HTTP_CONNECTION)).getMetrics();
-        metrics.reset();
+    private static final String DIGEST_PARAMETERS = DigestParameters.VARIABLE_NAME;
+
+    
+    private static final HttpRequestInterceptor PREEMPTIVE_AUTH_INTERCEPTOR = new HttpRequestInterceptor() {
```
Comment (ham1) [inline src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPHC4Impl.java]: This is almost 100 lines and with lots of nesting becomes very hard to read and review, could it be split into smaller methods?

**Review by ham1 — COMMENTED** [review]
Just a few minor comments, hopefully others have time to provide more detailed feedback.

**Comment by pmouawad** [comment]
Merged:

- http://svn.apache.org/viewvc?rev=1825269&view=rev

**Pull request closed without merging**